packages/hoppscotch-common/src/components/http/Codegen.vue

```diff
@@ -202,6 +202,27 @@ const getCurrentValue = (env: AggregateEnvironment) => {
   )?.currentValue
 }
 
+const getFinalURL = (url: string) => {
+  // If the URL is empty, return "https://"
+  // This is to ensure that the URL is always valid and can be used in code generation
+  if (!url) return "https://"
+
+  // If the URL is not empty, trim it and check if it starts with http:// or https://
+  // If it does not, check if it starts with <<, if it does not,
+  // then check if it is a valid domain or localhost, and prepend http:// or https accordingly
+  const endpoint = url.trim()
+  if (!/^http[s]?:\/\//.test(endpoint) && !endpoint.startsWith("<<")) {
+    const domain = endpoint.split(/[/:#?]+/)[0]
+    if (domain === "localhost" || /([0-9]+\.)*[0-9]/.test(domain)) {
+      url = "http://" + endpoint
+    } else {
+      url = "https://" + endpoint
+    }
+  }
+
+  return url
+}
```

Comment:
We'd also need to account for initial and trailing whitespaces as part of the endpoint. How about the below implementation?

```ts
const getFinalURL = (input: string): string => {
  if (!input) {
    return "https://"
  }

  let url = input.trim()

  // Fix malformed protocols
  url = url.replace(/^https?:\s*\/+\s*/i, (match) =>
    match.toLowerCase().startsWith("https") ? "https://" : "http://"
  )

  if (!/^https?:\/\//i.test(url) && !url.startsWith("<<")) {
    const endpoint = url
    const domain = endpoint.split(/[/:#?]+/)[0]

    const isLocalOrIP = /^(localhost|(\d{1,3}\.){3}\d{1,3})$/.test(domain)
    url = (isLocalOrIP ? "http://" : "https://") + endpoint
  }

  return url
}
```